x-pack/plugins/ml/public/application/components/custom_urls/custom_url_editor/custom_time_range_picker.tsx

```diff
@@ -0,0 +1,159 @@
+  onCustomTimeRangeChange: (customTimeRange?: { start: Moment; end: Moment }) => void;
+  customTimeRange?: { start: Moment; end: Moment };
+}
+
+/*
+ * React component for the form for adding a custom time range.
+ */
+export const CustomTimeRangePicker: FC<CustomUrlTimeRangePickerProps> = ({
+  onCustomTimeRangeChange,
+  customTimeRange,
+}) => {
+  const [showCustomTimeRangeSelector, setShowCustomTimeRangeSelectorked1] =
+    useState<boolean>(false);
+  const {
+    services: {
+      data: {
+        query: {
+          timefilter: { timefilter },
+        },
+      },
+    },
+  } = useMlKibana();
+
+  const onCustomTimeRangeSwitchChange = (e: {
+    target: { checked: React.SetStateAction<boolean> };
+  }) => {
+    if (e.target.checked === false) {
+      // Clear the custom time range so it isn't persisted
+      onCustomTimeRangeChange(undefined);
+    }
+    setShowCustomTimeRangeSelectorked1(e.target.checked);
+  };
+
+  // If the custom time range is not set, default to the timefilter settings
+  const currentTimeRange = useMemo(
+    () =>
+      customTimeRange ?? {
+        start: moment(timefilter.getAbsoluteTime().from),
+        end: moment(timefilter.getAbsoluteTime().to),
+      },
+    [customTimeRange, timefilter]
+  );
+
+  const handleStartChange = (date: moment.Moment) => {
+    onCustomTimeRangeChange({ ...currentTimeRange, start: date });
+  };
+  const handleEndChange = (date: moment.Moment) => {
+    onCustomTimeRangeChange({ ...currentTimeRange, end: date });
+  };
+
+  const { start, end } = currentTimeRange;
+
+  return (
+    <>
+      <EuiSpacer size="s" />
+      <EuiFlexGroup gutterSize={'none'}>
+        <EuiFlexItem grow={false}>
+          <EuiIconTip
+            content={i18n.translate('xpack.ml.customUrlsEditor.customTimeRangeTooltip', {
+              defaultMessage: 'If not set, time range defaults to global settings.',
+            })}
+            position="top"
+            type="iInCircle"
+          />
+        </EuiFlexItem>
+        <EuiFlexItem>
+          <EuiFormRow
+            display="columnCompressedSwitch"
+            label={
+              <FormattedMessage
+                id="xpack.ml.customUrlsEditor.customTimeRangeSwitch"
+                defaultMessage="Add custom time range?"
+              />
+            }
+          >
+            <EuiSwitch
+              showLabel={false}
+              label={
+                <FormattedMessage
+                  id="xpack.ml.customUrlsEditor.addCustomTimeRangeSwitchLabel"
+                  defaultMessage="Add custom time range switch"
+                />
+              }
+              checked={showCustomTimeRangeSelector}
+              onChange={onCustomTimeRangeSwitchChange}
+              compressed
+            />
+          </EuiFormRow>
+        </EuiFlexItem>
+      </EuiFlexGroup>
+      {showCustomTimeRangeSelector ? (
+        <>
+          <EuiSpacer size="s" />
+          <EuiFormRow
```

Comment:
If the destination data view of the job has a timestamp field, and the target page is either Dashboard or Discover with a data view that has a timestamp field, it would be good to include an extra option here as well as the custom time range option, where the user can specify an interval. The action here would be to open the target page showing `interval` either side of the time of the row the action is being run on, similar to the option we have for anomaly detection job custom URLs:

<img width="482" alt="image" src="https://user-images.githubusercontent.com/7405507/233434509-c6e92720-f484-4864-80dc-e71f5a0bec53.png">

Reply:
This item is part of the overall meta issue https://github.com/elastic/kibana/issues/150375 and can be done in a separate PR 👍